File: 21551-naveen-siva-kumar/week7/148-sort-list/148-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if(head == NULL || head ->next == NULL)
            return head;
        
        ListNode* sentinal = new ListNode(0,head);
        ListNode* fast = head;
        ListNode* slow = head;
        ListNode* temp;
        
        while(fast != NULL && fast->next!=NULL){
            
            temp=slow;
            slow = slow->next;
            fast=fast->next->next;
            
        }
        temp->next = NULL;
        
        ListNode* left = sortList(head);
        ListNode* right = sortList(slow);
        
        return merge(left,right);
        
    }
    ListNode* merge(ListNode* l,ListNode* r){
    
        ListNode *sentinal = new ListNode(0);
        ListNode *curr = sentinal;
        
        while(l != NULL && r != NULL){
            
            if(l->val <= r->val){
                curr -> next = l;
                l = l -> next;
            }
            
            else{
                curr -> next = r;
                r = r -> next;
            }
        
            curr = curr->next;
        
        }
        
        if(l != NULL){
            curr -> next = l;
            l = l->next;
        }
        
        if(r != NULL){
            curr -> next = r;
            r = r ->next;
        }
        
        return sentinal->next;

    
    }
};

File: 21557 - Mukesh J/Week-7/148-sort-list/148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if(head==null || head.next==null)
            return head;
        ListNode slow= head, fast=head;
        ListNode prev=null;
        while(fast!=null && fast.next!=null )
        {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        prev.next = null;
        
        ListNode left= sortList(head);
        ListNode right= sortList(slow);
        
        return merge(left,right);
    }
    
    public ListNode merge(ListNode left, ListNode right)
    {
        ListNode dummy= new ListNode(0), temp = dummy;
        while(left!=null && right!=null)
        {
            if(left.val<right.val)
            {
                temp.next=left;
                left=left.next;
            }
            else
            {
                temp.next=right;
                right=right.next;
            }
            temp= temp.next;
        }
        if(left!=null)
            temp.next=left;
        if(right!=null)
            temp.next=right;
        return dummy.next;
    }
}